File: 6- Reinforcement learning/Thompson sampling/R/thompson_sampling.R
#RE-INFORCEMENT LEARNING
#THOMPSON SAMPLING 

#import the dataset
dataset = read.csv('Ads_CTR_Optimisation.csv')

#Implement T.S.
N = 10000
d = 10
numbers_of_reward_1 = integer(d)
numbers_of_reward_0 = integer(d)
ads_selected = integer(0)
total_reward = 0 
#loop over all users
for(n in 1:N){
  ad = 0 
  max_random =0
  #loop over all versions of the ads
  for (i in 1:d){
    random_beta = rbeta(n= 1, 
                        shape1= numbers_of_reward_1[i] + 1, 
                        shape2 = numbers_of_reward_0[i]+1)
    
    if(random_beta > max_random){
      max_random = random_beta
      ad = i 
    }
  }
  ads_selected = append(ads_selected,ad)
  reward = dataset [n, ad]
  if(reward ==1)
  {
    numbers_of_reward_1[ad] = numbers_of_reward_1[ad]+1
  }else{
    numbers_of_reward_0[ad] = numbers_of_reward_0[ad]+1
    
  }
}


#visualizing results with an histogram
hist(ads_selected, col= 'blue', main = 'Histogram of ads selection', xlab = 'Ads', ylab = '# of time each time was selected')

File: 6- Reinforcement learning/Thompson sampling/python/thompson_sampling.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Apr  9 18:50:11 2017

THOMPSON SAMPLING

@author: Ilaria
"""

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import random

dataset = pd.read_csv('Ads_CTR_Optimisation.csv')


#implementing UCB
N = 10000
d = 10
numbers_of_rewards_1 = [0]*d
numbers_of_rewards_0 = [0]*d
ads_selected = []    
total_reward = 0             
#all users                  
for n in range(0,N):
    ad = 0
    max_random = 0
    #all version of the ads
    for i in range(0,d):
        random_beta = random.betavariate(numbers_of_rewards_1[i]+1, numbers_of_rewards_0[i]+1); 
        
        if random_beta > max_random:
            max_random = random_beta
            ad = i
    ads_selected.append(ad)
    reward = dataset.values[n, ad]
    if(reward ==1 ):
        numbers_of_rewards_1[ad] = numbers_of_rewards_1[ad] +1
    else:
        numbers_of_rewards_0[ad] = numbers_of_rewards_0[ad] +1
    total_reward = total_reward + reward
    
    
#visualizing the results
plt.hist(ads_selected)
plt.title('Histogram of ads selection')
plt.xlabel('ads')
plt.ylabel('Number of times each ad was selected')    
    
    